frame/support/procedural/src/construct_runtime/mod.rs

```diff
@@ -346,33 +347,57 @@ fn decl_all_pallets<'a>(
 		types.extend(type_decl);
 
 		if pallet_declaration.cfg_pattern.is_empty() {
-			for names in names_by_feature.values_mut() {
+			for (_, names) in features_to_names.iter_mut() {
 				names.push(&pallet_declaration.name);
 			}
 		} else {
-			for (feature_set, names) in &mut names_by_feature {
+			for (feature_set, names) in &mut features_to_names {
 				// Rust tidbit: if we have multiple `#[cfg]` feature on the same item, then the
 				// predicates listed in all `#[cfg]` attributes are effectively joined by `and()`,
 				// meaning that all of them must match in order to activate the item
 				let is_feature_active = pallet_declaration.cfg_pattern.iter().all(|expr| {
 					expr.eval(|pred| match pred {
-						Predicate::Feature(f) => feature_set.contains(&f),
-						Predicate::Test => feature_set.contains(&&"test"),
+						Predicate::Feature(f) => feature_set.contains(f),
+						Predicate::Test => feature_set.contains(&"test"),
 						_ => false,
 					})
 				});
+
 				if is_feature_active {
 					names.push(&pallet_declaration.name);
 				}
 			}
 		}
 	}
 
-	let all_pallets_without_system = names_by_feature.iter().map(|(feature_set, names)| {
-		let mut feature_set = feature_set.iter().collect::<HashSet<_>>();
-		let test_cfg = feature_set.remove(&&&"test").then_some(quote!(test)).into_iter();
-		let feature_set = feature_set.into_iter();
-		let attr = quote!(#[cfg(all( #(#test_cfg),* #(feature = #feature_set),* ))]);
+	// All possible features. This will be used below for the empty feature set.
+	let mut all_features = features_to_names
+		.iter()
+		.flat_map(|f| f.0.iter().cloned())
+		.collect::<HashSet<_>>();
+	let attribute_to_names = features_to_names
+		.into_iter()
+		.map(|(mut features, names)| {
+			// If this is the empty feature set, it needs to be changed to negate all available
+			// features. So, we ensure that there is some type declared when all features are not
+			// enabled.
+			if features.is_empty() {
+				let test_cfg = all_features.remove("test").then_some(quote!(test)).into_iter();
+				let features = all_features.iter();
+				let attr = quote!(#[cfg(all( #(not(#test_cfg)),* #(not(feature = #features)),* ))]);
```

Comment:
Yeah this is exactly the line that I was missing... I told myself to include this while coding, but I guess it just slipped through the cracks. Thanks!